src/device_memory_tracker.cc

```diff
@@ -0,0 +1,288 @@
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION));
+    break;
+  }
+  case CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED: {
+    std::cerr << "CUPTI has been initialized elsewhere, assuming the implementation is the same" << std::endl;
+    break;
+  }
+  default:
+    // other error, should propagate and disable memory tracking for the backend
+    throw std::runtime_error("Unexpected failure on configuring CUPTI.");
+  }
+}
+
+int
+DeviceMemoryTracker::CudaDeviceCount()
+{
+  if (tracker_) {
+    return tracker_->device_cnt_;
+  }
+  return 0;
+}
+
+bool
+DeviceMemoryTracker::Init()
+{
+  if (tracker_ == nullptr) {
+    try {
+      tracker_.reset(new DeviceMemoryTracker());
+    } catch (const std::runtime_error& ex) {
+      // Fail initialization 
+      return false;
+    }
+  }
+  return true;
+}
+
+void
+DeviceMemoryTracker::TrackThreadMemoryUsage(MemoryUsage* usage)
+{
+  if (tracker_) {
+    CUPTI_CALL(cuptiActivityPushExternalCorrelationId(
+        CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN,
+        reinterpret_cast<uint64_t>(&usage->cupti_tracker_)));
+    usage->tracked_ = true;
+  } else {
+    throw std::runtime_error("DeviceMeoryTracker::Init() must be called before using any DeviceMeoryTracker features.");
+  }
+}
+
+void
+DeviceMemoryTracker::UntrackThreadMemoryUsage(MemoryUsage* usage)
+{
+  if (tracker_) {
+    uint64_t id;
+    CUPTI_CALL(cuptiActivityPopExternalCorrelationId(
+        CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN, &id));
+    CUPTI_CALL(cuptiActivityFlushAll(0));
+    usage->tracked_ = false;
+  } else {
+    throw std::runtime_error("DeviceMeoryTracker::Init() must be called before using any DeviceMeoryTracker features.");
+  }
+}
+
+void
+DeviceMemoryTracker::TrackActivityInternal(CUpti_Activity* record)
+{
+  switch (record->kind) {
+    case CUPTI_ACTIVITY_KIND_MEMORY2: {
+      CUpti_ActivityMemory3* memory_record = (CUpti_ActivityMemory3*)record;
+      TRITONBACKEND_CuptiTracker* usage = nullptr;
+      {
+        std::lock_guard<std::mutex> lk(mtx_);
+        auto it = activity_to_memory_usage_.find(memory_record->correlationId);
+        if (it != activity_to_memory_usage_.end()) {
+          usage = reinterpret_cast<TRITONBACKEND_CuptiTracker*>(it->second);
+          activity_to_memory_usage_.erase(it);
+        }
+      }
+      // Ignore memory record that is not associated with a TRITONBACKEND_CuptiTracker
+      // object
+      if (usage != nullptr) {
+        const bool is_allocation =
+            (memory_record->memoryOperationType ==
+             CUPTI_ACTIVITY_MEMORY_OPERATION_TYPE_ALLOCATION);
+        const bool is_release =
+            (memory_record->memoryOperationType ==
+             CUPTI_ACTIVITY_MEMORY_OPERATION_TYPE_RELEASE);
+        if (is_allocation || is_release) {
+          switch (memory_record->memoryKind) {
+            case CUPTI_ACTIVITY_MEMORY_KIND_DEVICE: {
+              if (memory_record->deviceId >= usage->cuda_array_len_) {
```

Comment:
since the logic here is the same, - could refactor to an inline function taking the array and arraylen and memory_record.

Another option could be to set local variables (i.e. auto memory_usage_array = usage->cuda_byte_size, memory_usage_array_len = usage->cuda_array_len.